torch/csrc/distributed/c10d/Functional.cpp

```diff
@@ -0,0 +1,285 @@
+}
+
+at::Tensor allocateReduceScatterOutput(
+    const at::Tensor& input,
+    int64_t groupSize) {
+  auto outputSize = input.sizes().vec();
+  if (outputSize[0] % groupSize != 0) {
+    LOG(WARNING) << "The first dimension of the reduce_scatter input ("
+                 << outputSize[0] << ") is not divisible by the group size ("
+                 << groupSize << ").";
+  }
+  outputSize[0] /= groupSize;
+  return at::empty(
+      outputSize,
+      at::TensorOptions().dtype(input.dtype()).device(input.device()));
+}
+
+std::vector<at::Tensor> reduce_scatter_tensor_coalesced(
+    const std::vector<at::Tensor>& inputs,
+    const std::string& reduceOp,
+    const std::string& tag,
+    const std::vector<int64_t>& ranks,
+    const int64_t groupSize) {
+  c10d::ReduceScatterOptions opts;
+  auto reduceOpIt = strToReduceOp.find(reduceOp);
+  if (reduceOpIt == strToReduceOp.end()) {
+    LOG(FATAL) << "Unrecognized reduceOp: " << reduceOp;
+  }
+  opts.reduceOp = reduceOpIt->second;
+
+  std::vector<at::Tensor> outputs;
+  for (const auto& tensor : inputs) {
+    outputs.push_back(allocateReduceScatterOutput(tensor, groupSize));
+  }
+  auto pg = c10d::ProcessGroup::resolveFromName(tag);
+  if (groupSize != pg->getSize()) {
+    LOG(FATAL) << "Mismatch between group size argument and world size"
+                  "of process group ("
+               << tag << ")";
+  }
+  auto work = pg->reduce_scatter_tensor_coalesced(
+      outputs, const_cast<std::vector<at::Tensor>&>(inputs), opts);
+  for (const auto& tensor : outputs) {
+    // c10d::Work::wait is idempotent
+    pendingWork[tensor.data_ptr()] = work;
+  }
+  return outputs;
+}
+
+at::Tensor reduce_scatter_tensor(
+    at::Tensor input,
+    const std::string& reduceOp,
+    const std::string& tag,
+    const std::vector<int64_t>& ranks,
+    const int64_t groupSize) {
+  std::vector<at::Tensor> inputs{input};
+  return reduce_scatter_tensor_coalesced(
+      inputs, reduceOp, tag, ranks, groupSize)[0];
+}
+
+at::Tensor wait_tensor(const at::Tensor& tensor) {
+  auto it = pendingWork.find(tensor.data_ptr());
+  if (it == pendingWork.end()) {
+    LOG(FATAL)
+        << "No pending collective is associated with the input tensor. "
+           "This typically means that the input tensor is not a collective output, "
+           "or the tensor has already been waited on.";
+  }
+  it->second->wait();
+  pendingWork.erase(it);
+  return tensor;
+}
+
+} // namespace
+
+TORCH_LIBRARY(_c10d_functional, m) {
+  // TODO(yifu): is it neccessary to keep the same argument names used in
+  // c10d_functional? It would be nice to modify it to be more consistent.
+  m.def(
+      "all_reduce(Tensor self, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor",
+      torch::dispatch(
+          c10::DispatchKey::CompositeExplicitAutograd, ::all_reduce));
+
+  m.def(
+      "all_reduce_(Tensor(a!) self, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor",
```

Comment:
all need to tag output. 

see examples in native_functions.yaml
e.g.
`- func: normal_(Tensor(a!) self, float mean=0, float std=1, *, Generator? generator=None) -> Tensor(a!)`

Reply:
also see https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/README.md